security/pkg/pki/util/keycertbundle.go

```diff
@@ -233,6 +233,24 @@ func (b *KeyCertBundleImpl) ExtractCACertExpiryTimestamp() (float64, error) {
 	return extractCertExpiryTimestamp("CA cert", b.GetCertChainPem())
 }
 
+// TimeBeforeNextRotate returns time duration to the next cert rotation
+func TimeBeforeNextRotate(certBytes []byte, workloadTTL time.Duration, now time.Time) (time.Duration, error) {
+	if len(certBytes) == 0 {
```

Comment:
Can you add a comment explaining why the duration to the next cert rotation is workloadTTL when certBytes is empty?

Reply:
MinTTL() is only called in the line 290 as follows:
certTTL, err := util.MinTTL(ca.keyCertBundle.GetCertChainPem(), opts.DefaultCertTTL, time.Now())

When certBytes is empty, should MinTTL() return an error?

Reply:
I thought that `certBytes` was not empty because of this piece of code:

`pilot/pkg/bootstrap/istio_ca.go:360`
```
caOpts, err = ca.NewPluggedCertIstioCAOptions(certChainFile, signingCertFile, signingKeyFile,
			rootCertFile, workloadCertTTL.Get(), maxCertTTL, opts.Namespace, client)
```

it seems that inside this function is inserted the `KeyCertBundle` with the `certChain`

Am I missing something?

Reply:
If certBytes is not supposed to be empty, MinTTL() should return an error instead of returning nil (i.e., no error) when certBytes is empty.

Reply:
This is needed because we are passing the intermediate cert as the `certBytes`, which it could not exist (I think)

Reply:
When the intermediate cert is passed as the certBytes and it is empty, should MinTTL() return minimum(defaultTTL, TTL-of-root-cert), instead of defaultTTL?

Reply:
let me see if I get you, you want this return `defaultTTL` (when it's intermediate cert) out of the `MinTTL` function?

then I can return error in `MinTTL` if it's empty

is that right?

Reply:
I am not sure what you mean by "you want return defaultTTL (when it's intermediate cert) out of the MinTTL function". 

I think the semantic of MinTTL() function should be returning the minimum TTL of the leaf cert, the intermediate certs, and the root cert.

Reply:
I see your point, I'm definitely not against it, but to accomplish that I'll have to move this logic (return defaultTTL if there is no CA certificate) to the `NewIstioCA` function.

that is ok or you see a different way to do it?

Reply:
For MinTTL() function:
- MinTTL() function should be returning the minimum TTL of the leaf cert, the intermediate CA certs, and the root CA cert.
- When there are no intermediate CA certs and the root CA cert, MinTTL() return an error since there is no CA certs.
- When there are no intermediate CA certs but there is root CA cert, MinTTL() return the minimum TTL of the leaf cert and the root CA cert.
- When there are intermediate CA certs but there is no root CA cert, MinTTL() return the minimum TTL of the leaf cert and intermediate CA certs.

Reply:
So it's better to call this function by another name (or create another implementation) because I don't need any of those values suggested.

The only thing that I need is to return the min value between workloadTTL and intermediate cert TTL.

Reply:
I've created a more generic function (`TimeBeforeCertExpires`), and the logic about using cert chain TTL or workload TTL is in `ca.go` now